frontend/utility/PreviewProgramSizeObserver.cpp

```diff
@@ -0,0 +1,221 @@
+
+		if (!parent1 || !parent2) {
+			break;
+		}
+
+		// Found sibling containers
+		if (parent1 == parent2) {
+			return {ancestor1, ancestor2};
+		}
+
+		if (parent1->isAncestorOf(parent2)) {
+			ancestor2 = parent2;
+		} else if (parent2->isAncestorOf(parent1)) {
+			ancestor1 = parent1;
+		} else {
+			ancestor1 = parent1;
+			ancestor2 = parent2;
+		}
+	}
+
+	return {a, b};
+}
+
+void PreviewProgramSizeObserver::syncContainerSizes(int containerSizeDelta)
+{
+	auto setMax = (orientation == Qt::Horizontal) ? [](QWidget* widget, int value) {
+		widget->setMaximumWidth(value);
+		} : [](QWidget* widget, int value) {
+		widget->setMaximumHeight(value);
+		};
+
+	QLayout *ancestorLayout = ancestorContainer->layout();
+	if (QBoxLayout *boxLayout = qobject_cast<QBoxLayout *>(ancestorLayout)) {
+		setOrientation((boxLayout->direction() == QBoxLayout::LeftToRight) ? Qt::Horizontal : Qt::Vertical);
+	}
+
+	if (orientation == Qt::Horizontal) {
+		leftContainer->setMaximumHeight(leftOriginalMaxSize.height());
+		rightContainer->setMaximumHeight(rightOriginalMaxSize.height());
+	} else {
+		leftContainer->setMaximumWidth(leftOriginalMaxSize.width());
+		rightContainer->setMaximumWidth(rightOriginalMaxSize.width());
+	}
+
+	int leftInner = (orientation == Qt::Horizontal) ? left->width() : left->height();
+	int rightInner = (orientation == Qt::Horizontal) ? right->width() : right->height();
+
+	int leftOuter = (orientation == Qt::Horizontal) ? leftContainer->width() : leftContainer->height();
+	int rightOuter = (orientation == Qt::Horizontal) ? rightContainer->width() : rightContainer->height();
+
+	int totalOuter = leftOuter + rightOuter;
+	if (containerSizeDelta >= 0) {
+		totalOuter += containerSizeDelta;
+	}
+
+	int leftOffset = leftOuter - leftInner;
+	int rightOffset = rightOuter - rightInner;
+
+	int targetInner = (totalOuter - leftOffset - rightOffset) / 2;
+
+	leftTargetSize = targetInner + leftOffset;
+	rightTargetSize = targetInner + rightOffset;
+
+	if (containerSizeDelta >= 0) {
+		setMax(leftContainer, leftTargetSize);
+		setMax(rightContainer, rightTargetSize);
+	} else {
+		// Container shrunk, only set max size on larger widget
+		if (leftInner > rightInner) {
+			setMax(leftContainer, leftTargetSize);
+			setMax(rightContainer, QWIDGETSIZE_MAX);
+		} else {
+			setMax(leftContainer, QWIDGETSIZE_MAX);
+			setMax(rightContainer, rightTargetSize);
+		}
+
+		// Force a second recalculation
+		QTimer::singleShot(1, this, [&, setMax]() {
+			if (updating) {
+				return;
+			}
+
+			updating = true;
+
+			setMax(leftContainer, leftTargetSize);
+			setMax(rightContainer, rightTargetSize);
+
+			updating = false;
+		});
+	}
+}
+
+bool PreviewProgramSizeObserver::eventFilter(QObject *target, QEvent *event)
+{
+	if (event->type() != QEvent::Resize && event->type() != QEvent::LayoutRequest) {
+		return QObject::eventFilter(target, event);
+	}
+
+	if (!left || !right || !leftContainer || !rightContainer) {
+		deleteLater();
```

Comment:
Is it canonical in Qt to schedule the deallocation of an object in reaction to an `eventFilter` call?

As far as I can tell you add it as a child element to a parent widget, so any cleanup should happen as part of that lifetime management anyway?

Reply:
The eventFilter itself is tied to this object. This schedules the object for deletion if any of the relevant widget pointers are no longer valid for some reason.

deleteLater schedules for removal when it is safe for the event loop to do so.